src/main/java/it/rate/webapp/services/RatingService.java

```diff
@@ -2,82 +2,93 @@
-                  ratingRepository.findById(
-                      new RatingId(loggedUser.getId(), place.getId(), criterion.getId()));
-              if (optRating.isPresent()) {
-                if (value == null) {
-                  ratingRepository.deleteById(
-                      new RatingId(loggedUser.getId(), place.getId(), criterion.getId()));
-                  return;
-                }
-                Rating existingRating = optRating.get();
-                existingRating.setScore(value);
-                ratingRepository.save(existingRating);
-              } else {
-                if (value == null) {
-                  return;
-                }
-                Rating newRating = new Rating(loggedUser, place, criterion, value);
-                ratingRepository.save(newRating);
-              }
+              Criterion criterion =
+                  ratedCriteria.stream()
+                      .filter(c -> Objects.equals(c.getId(), key))
+                      .findAny()
+                      .orElseThrow(
+                          () ->
+                              new ResponseStatusException(
+                                  HttpStatus.BAD_REQUEST, "Invalid criterion in ratings"));
+              updateOrCreateRating(appUser, place, criterion, value);
             });
   }
 
-  private AppUser getLoggedUser(Principal principal) {
-    if (principal == null) {
-      throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "User not authenticated");
+  private void updateOrCreateRating(AppUser appUser, Place place, Criterion criterion, Integer value) {
+    RatingId ratingId = new RatingId(appUser.getId(), place.getId(), criterion.getId());
+    Optional<Rating> optRating = ratingRepository.findById(ratingId);
+
+    if (value == null) {
+      if (optRating.isPresent()) {
+        ratingRepository.deleteById(ratingId);
+      }
+      return;
+    }
+
+    if (optRating.isPresent()) {
+      Rating existingRating = optRating.get();
+      existingRating.setScore(value);
+      ratingRepository.save(existingRating);
+    } else {
+      Rating newRating = new Rating(appUser, place, criterion, value);
+      ratingRepository.save(newRating);
     }
-    return userService.getByEmail(principal.getName());
   }
 
-  private Place getPlace(Long placeId) {
-    Optional<Place> optPlace = placeService.findById(placeId);
-    if (optPlace.isEmpty()) {
-      throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Invalid argument");
+  private Set<Criterion> validateRatings(RatingsDTO ratings, Place place) {
+    List<Criterion> placeCriteria = place.getInterest().getCriteria();
+    Set<Criterion> ratedCriteria = new HashSet<>();
+
+    ratings
+        .ratings()
+        .forEach(
+            (key, value) -> {
+              Criterion criterion = getCriterion(key);
+              ratedCriteria.add(criterion);
+              if (value != null && (value < 1 || value > 10)) {
```

Comment:
It is also possible to use the @Min and @Max annotation like this in the RatingsDTO:


```
public record RatingsDTO(Map<Long, @Min(value = 1) @Max(value = 10) Integer> ratings) {
```

and then validate like this:
```

@Autowired private Validator validator;
....

if (!validator.validate(ratings).isEmpty()) {
  throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Invalid rating value");
}
```

You don't have to change this, it's just good to know and could save some time if you ever need to do more than just one simple validation.